feat: Automatic sqlite-database-integration upgrade

dcalhoun opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "Fixes" keyword and use "Related to" instead.
-->

Resolves https://github.com/Automattic/dotcom-forge/issues/6939.
Fixes https://github.com/Automattic/dotcom-forge/issues/6922. 
Relates to https://github.com/Automattic/dotcom-forge/issues/6598. 
Relates to https://github.com/Automattic/studio/issues/132.
Relates to https://github.com/Automattic/local-environment/pull/245 and https://github.com/Automattic/local-environment/pull/228. 

## Proposed Changes

- Update the `sqlite-databse-integration` installation to the **latest release tag** on app start. 
- Update the `sqlite-databse-integration` installation to the **latest release tag** on site server start. 

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

> [!IMPORTANT]
> In order to observe the `sqlite-integration-plugin` upgrade, you must first install a legacy version of the plugin.
>
> 1. Check out the `trunk` branch.
> 1. Download a legacy version of `sqlite-integration-plugin` — e.g., [v2.1.9](https://github.com/WordPress/sqlite-database-integration/releases/tag/v2.1.9).
> 1. "Install" the legacy version by replacing the existing copy found in the following locations. 
>
>    **Note:** naming the directory `sqlite-database-integration-main` (legacy name) would simulate testing the rollout of these changes to existing installs/sites; naming the directory `sqlite-database-integration` (new name) would simulate testing an additional SQLite upgrade (second, third, etc) after a previous install (first SQLite upgrade, new site creation using these changes, etc).
>    - **Bundled Installation:** This is found in the cloned repository at `wp-files/sqlite-database-integration-main`.
>    - **Downloaded Installation:** On macOS, this is found in `~/Library/Application\ Support/Studio/server-files/sqlite-database-integration-main`. 
> 1. Start the app server. (Note, you should not run `npm install` at this point as it would update the legacy bundled installation.)
> 1. Create a new site ("Legacy Site").
> 1. Verify the Legacy Site's `wp-content/mu-plugins` directory contains the legacy version of the plugin.

**1: Site actions remain stable without a network connection**
1. Check out the proposed changes.
1. Run `npm install` to download/bundle the latest version of `sqlite-database-integration`.
1. Disable your internet connection. 
1. Start the app server.
1. Create a new site.
1. Verify no errors are thrown during site creation or site server start.

**2: New sites leverage the latest version**
1. Check out the proposed changes.
1. Start the app server.
1. Create a new site. 
1. Verify the `sqlite-integration-plugin` version found in the site's `wp-content/mu-plugins` directory is the [latest release](https://github.com/WordPress/sqlite-database-integration/releases).

**3: Existing sites upgrade when started**
1. Check out the proposed changes.
1. Start the app server.
1. Start the Legacy Site's server.
1. Verify the `sqlite-integration-plugin` version found in the site's `wp-content/mu-plugins` directory is the [latest release](https://github.com/WordPress/sqlite-database-integration/releases).

**4: Woocommerce loads without database query errors**

1. Check out the proposed changes.
1. Start the app server.
1. Install and activate the WooCommerce plugin (Steps outlined in https://github.com/Automattic/dotcom-forge/issues/6922).
1. Verify no database query errors are logged and the plugin page loads successfully. **Note:** If MailPoet is installed and activated, [errors may display](https://github.com/Automattic/studio/issues/132#issuecomment-2120492613). 

**~5: UpdraftPlus loads without database query errors~**

Test case removed as the proposed changes do not address [all errors](https://github.com/WordPress/wordpress-playground/issues/1272#issuecomment-2120820926). 

**6: Fresh Studio installs successfully create and start sites**
1. Delete all existing Studio sites. 
1. Clear out the bundled and downloaded `sqlite-database-integration` installations. 
1. Run `npm install` to download/bundle the latest version of `sqlite-database-integration`.
1. Start the app server. 
1. Create a new site. 
1. Verify the `sqlite-integration-plugin` version found in the site's `wp-content/mu-plugins` directory is the [latest release](https://github.com/WordPress/sqlite-database-integration/releases).

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Have you checked for TypeScript, React or other console errors?

dcalhoun committed:
feat: Starting the app updates sqlite-database-integration installation

Ensure new sites receive the latest sqlite-database-integration fixes
and improvements.

dcalhoun committed:
feat: Download latest sqlite-databse-integration release tag

Improve stability and enable version comparisons for future upgrades.

dcalhoun committed:
feat: Starting a site updates an outdated sqlite-integration-plugin

Ensure existing sites receive the latest fixes and features.

dcalhoun committed:
fix: Avoid unnecessary sqlite-database-integration upgrades

If version comparison fails, assume the installed version is the latest.
For example, fetching the latest version while offline will result in a
comparison failure.

dcalhoun committed:
feat: Cache sqlite-database-integration versions for app session

Avoid repeatedly fetching the latest versions each time a site starts.

dcalhoun commented on scripts/download-wp-server-files.ts:
```diff
@@ -15,7 +15,7 @@ const FILES_TO_DOWNLOAD = [
 		name: 'sqlite',
 		description: 'SQLite files',
-		url: 'https://codeload.github.com/WordPress/sqlite-database-integration/zip/refs/heads/main',
+		url: 'https://downloads.wordpress.org/plugin/sqlite-database-integration.zip',
```
Because the `sqlite-database-integration` GitHub [repository](https://github.com/WordPress/sqlite-database-integration/releases) names each release ZIP with the version number — e.g., `v2.1.10.zip` — [linking to the latest release](https://docs.github.com/en/repositories/releasing-projects-on-github/linking-to-releases#linking-to-the-latest-release) on GitHub does not work, as the ZIP file will always have a different name.

sejas replied:
I see we use wp.org for the bundled version and GitHub for the updates.
I would prefer if we download the plugin from the same source. Not a blocker, just a nice to have.

dcalhoun replied:
Agreed. I originally split the source due to a "need" to fetch versions, but, as you noted, that is not actually necessary. With the latest changes, the wp.org source is now used in all locations.

sejas replied:
Thank you for making the changes 🙌

dcalhoun commented on src/ipc-handlers.ts:
```diff
@@ -105,7 +106,7 @@ async function setupSqliteIntegration( path: string ) {
 		)
 	);
 	const sqlitePluginPath = nodePath.join( wpContentPath, 'mu-plugins', SQLITE_FILENAME );
-	await copySync( nodePath.join( getServerFilesPath(), SQLITE_FILENAME ), sqlitePluginPath );
```
Awaiting this synchronous method has no impact.

sejas replied:
Nice catch!, I've seen that copySync returns `void` 👍 .

dcalhoun commented on src/lib/sqlite-versions.ts:
```diff
@@ -0,0 +1,74 @@
+
+	try {
+		const response = await fetch(
+			'https://api.github.com/repos/WordPress/sqlite-database-integration/releases'
```
The WordPress plugin directory lacks a list of [past versions](https://api.wordpress.org/plugins/info/1.2/?action=plugin_information&slug=sqlite-database-integration) for the `sqlite-database-integration` plugin. So, this relies upon GitHub repository releases instead.

sejas replied:
Out of curiosity, why do we need the list of past versions?
I'm seeing this fetch is only called on `getLatestSqliteVersion`, which returns only the latest version.

dcalhoun replied:
Good question. We do not. This unnecessary complexity originated from too closely mirroring the `wp-versions` logic. I simplified this `sql-versions` logic, removing the usage of a GitHub endpoint and references to versions other than the latest.

dcalhoun commented on vendor/wp-now/src/download.ts:
```diff
@@ -191,18 +191,26 @@ export async function downloadWordPress(
 	if (downloaded) {
-		fs.ensureDirSync(path.dirname(finalFolder));
-		fs.moveSync(tempFolder, finalFolder, {
+		const nestedFolder = SQLITE_FILENAME.replace(/main$/, version);
```
The resulting directory includes a tag suffix rather than the `-main` branch suffix. In an effort to [minimize changes](https://github.com/Automattic/studio/blob/14ef3c0f01effe46e48dd9a0b5f72f78ab0046f1/vendor/wp-now/src/constants.ts#L4) to `wp-now` source and simplify compatibility with existing installations, I opted to retain the `-main` suffix.

dcalhoun replied:
With the latest changes, this was simplified to merely remove the `-main` branch suffix that is not present when downloading a release from wp.org.

dcalhoun commented:
Expanded the testing instructions in the description as I realized this will also fix https://github.com/Automattic/dotcom-forge/issues/6922 and https://github.com/Automattic/dotcom-forge/issues/6598.

sejas commented on src/lib/sqlite-versions.ts:
```diff
@@ -0,0 +1,74 @@
+import path from 'path';
+import * as Sentry from '@sentry/electron/main';
```
I think we can remove the Sentry import from this file.

sejas reviewed (approved):
I tested all the scenarios and worked great in all of them.
I added a few comments. My only suggestion is to use the same source, wp.org or GitHub, for the bundled version and the start/updated version.


Updated existing SQLite version from `2.1.7` to `2.1.10`

https://github.com/Automattic/studio/assets/779993/aebfcf3b-a20e-4ebc-af09-fa01fa2470b1



https://github.com/Automattic/studio/assets/779993/e842afae-bb4b-4de4-a5bd-5da3222ff40e


https://github.com/Automattic/studio/assets/779993/e843daaa-f302-4f1f-bd79-d381746018a4


https://github.com/Automattic/studio/assets/779993/f9622d31-efee-4653-99b2-b150f0df27e1

dcalhoun committed:
refactor: Remove unused import

dcalhoun committed:
refactor: Remove unused references to specific past release versions

We only ever install the latest version, so we do not need the
unnecessary complexity of fetching past versions. Also, this allows us
to use the same download source the `sqlite-database-integration` plugin
throughout the source.

fluiddot commented on src/setup-wp-server-files.ts:
```diff
@@ -51,4 +52,5 @@ export default async function setupWPServerFiles() {
 	await copyBundledLatestWPVersion();
 	await copyBundledSqlite();
```
I noticed that after changes introduced in https://github.com/Automattic/studio/pull/136/commits/a73fe82af9a299942b53b5c6cee4081b0ab9312d, the bundled version of SQLite has the path `wp-files/sqlite-database-integration` (i.e. no longer contains the suffix `-main`). However, in `copyBundledSqlite` function, the source path used in the copy operation references `wp-files/sqlite-database-integration-main`. This results in not copying the SQLite plugin to the Studio's server files location, but creating an empty folder. Besides, having an empty folder affects the function `downloadSqliteIntegrationPlugin`, as it prevents the download from happening ([reference](https://github.com/Automattic/studio/blob/a73fe82af9a299942b53b5c6cee4081b0ab9312d/vendor/wp-now/src/download.ts#L108-L111)).

fluiddot replied:
Maybe we could simply rename [`SQLITE_FILENAME`](https://github.com/Automattic/studio/blob/a73fe82af9a299942b53b5c6cee4081b0ab9312d/vendor/wp-now/src/constants.ts#L4) to `sqlite-database-integration`, now that the `-main` suffix won't be included since we don't download the GitHub version.

dcalhoun replied:
> I noticed that after changes introduced in [a73fe82](https://github.com/Automattic/studio/commit/a73fe82af9a299942b53b5c6cee4081b0ab9312d), the bundled version of SQLite has the path `wp-files/sqlite-database-integration` (i.e. no longer contains the suffix `-main`). However, in `copyBundledSqlite` function, the source path used in the copy operation references `wp-files/sqlite-database-integration-main`. 

Thank you for testing and catching this. I'll address this oversight. 

> Maybe we could simply rename [`SQLITE_FILENAME`](https://github.com/Automattic/studio/blob/a73fe82af9a299942b53b5c6cee4081b0ab9312d/vendor/wp-now/src/constants.ts#L4) to `sqlite-database-integration`, now that the `-main` suffix won't be included since we don't download the GitHub version.

I considered this approach, but ultimately [opted to minimize](https://github.com/Automattic/studio/pull/136#discussion_r1604197174) `wp-now` changes. That said, I'm open to modifying the `SQLITE_FILENAME`. It would likely simplify the code. 👍🏻

fluiddot replied:
> I considered this approach, but ultimately [opted to minimize](https://github.com/Automattic/studio/pull/136#discussion_r1604197174) `wp-now` changes. That said, I'm open to modifying the `SQLITE_FILENAME`. It would likely simplify the code. 👍🏻

I updated locally the `SQLITE_FILENAME` constant and seems to work without further changes. @dcalhoun have you identified other parts that we might need to update in `wp-now`?

dcalhoun replied:
No, I had not explored the ramifications. Thank you for testing. I will modify the value and test myself as well.

dcalhoun replied:
This should be resolved now by https://github.com/Automattic/studio/pull/136/commits/f1d2dae74b3666c6a6af678b4d5c1d6fae99ca84. Thank you again for catching this. 🙇🏻

fluiddot commented:
> **Testing Instructions**
> UpdraftPlus loads without database query errors

I tested the plugin and I'm encountering the same JS exceptions outlined in https://github.com/Automattic/dotcom-forge/issues/6598:
```
ReferenceError: updraftlion is not defined
```

I presume that using the latest SQLite version solved database errors, but as far as I've tested, I think this PR won't solve the issue. The plugin doesn't seem to function properly, clicking the Backup Now button results in no actions.

dcalhoun commented:
> I tested the [UpdraftPlus] plugin and I'm encountering the same JS exceptions outlined in [Automattic/dotcom-forge#6598](https://github.com/Automattic/dotcom-forge/issues/6598):
> 
> ```
> ReferenceError: updraftlion is not defined
> ```

Thank you for sharing this. From earlier testing, my perception was that this error was present for non-Playground sites as well, but it appears I was mistaken. I'll debug this further.

dcalhoun committed:
refactor: Remove `-main` branch suffix from SQLite install

The `-main` suffix is no longer accurate now that we rely upon release
tags rather than downloading the latest development branch.

dcalhoun reviewed (commented):
Thank you @sejas and @fluiddot for reviewing and helping me test this thoroughly. Apologies for the overlooked cases. I believe this is ready for review again. 

1. I updated the test case 4 (WooCommerce) to note that [errors remain](https://github.com/Automattic/studio/issues/132#issuecomment-2120492613) if MailPoet is activated. 
2. I removed test case 5 (UpdraftPlus) as the proposed changes do not address [all the errors](https://github.com/WordPress/wordpress-playground/issues/1272#issuecomment-2120820926). 
3. I added test case 6 to ensure that fresh installations of Studio — which lack `sqlite-database-integration` plugin installations entirely — work as expected.

fluiddot commented on src/lib/sqlite-versions.ts:
```diff
@@ -0,0 +1,94 @@
+function getSqliteVersionFromInstallation( installationPath: string ): string {
+	let versionFileContent = '';
+	try {
+		versionFileContent = fs.readFileSync( path.join( installationPath, 'load.php' ), 'utf8' );
```
In general, we try to avoid using the sync version of FS functions to prevent blocking the main thread. I wonder if we could use the asynchronous one.

fluiddot reviewed (approved):
LGTM 🎊 ! I've tested the different test scenarios and worked as expected. Thanks @dcalhoun for adding this feature 🏅 !

I added a comment that shouldn't block the PR but might be interesting to consider.

_Nitpick:_ In the `Important` section of **Testing Instructions**, there are references to the old SQLite filename `sqlite-database-integration-main. I presume it should be `sqlite-database-integration`, without the suffix `-main`.

dcalhoun commented:
> _Nitpick:_ In the `Important` section of **Testing Instructions**, there are references to the old SQLite filename `sqlite-database-integration-main`. I presume it should be `sqlite-database-integration`, without the suffix `-main`.

@fluiddot Thanks for the note! 

My thought is that testing old-version SQLite installations with and without the `-main` suffix is relevant. The former applies for existing users receiving upgrades to existing installations with `-main`. The latter applies for users receiving upgrades to existing installations without `-main` (e.g., upgrading SQLite a second time or installing Studio for the first time after this change ships). 

I updated the documentation account for both of these nuanced circumstances.
